[Blocked] Add PCA.getAccount(oid) API to retrieve account by object id

heidijinxujia opened the pull request:
**Problem**
Currently MSAL getAccount expects a homeAccountIdentifier.  I think we should just an identifier.  This will allow for a simpler migration for customers who've already persisted the object id of the user in their system.

The oid is the obejct id, which should be the same value of local account id. For the accounts with same home accout id, the local account ids differ with tenants. 

**Solution**
Add the API to retrieve the account with oid
Remove the API to getAccount() with homeAccountId, because the aim here is to restrict the retrieving with oid only.

Also fixed the redirect uri bug founding during the E2E testing
- Update the redirectUri generation with the new format

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -360,6 +361,19 @@ public PublicClientApplicationConfiguration getConfiguration() {
+    /**
+     * Listener callback for asynchronous finding of IAccount matching the given oid.
+     */
+    public interface AccountLoadedCallback {
```
Can we re-tool the existing interface to accommodate returning 1 _or_ many accounts? I think we can keep the method name plural without confusing ppl

Something like
```java
public interface AccountsLoadedCallback<T> {
    void onAccountsLoaded(T account);
}
```

And then our various implementations will pass-in the right arg depending on what's needed
```java
new AccountsLoadedCallback<Account>() { // Just need one
    @Override
    public void onAccountsLoaded(Account account) {
        // Do something useful            
    }
}

new AccountsLoadedCallback<List<Account>>() { // Need multiple
    @Override
    public void onAccountsLoaded(List<Account> account) {
        // Something else! 
    }
}
```

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -554,51 +568,126 @@ public void run() {
+                                handler.post(new Runnable() {
+                                    @Override
+                                    public void run() {
+                                        callback.onAccountLoaded(AccountAdapter.adapt(accountRecord));
```
Certainly we can't be meaning to call this multiple times, right?

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -554,51 +568,126 @@ public void run() {
+                        "Found the matched account."
+                );
+
+                callback.onAccountLoaded(AccountAdapter.adapt(accountRecord));
```
Same here - invocation multiple times?

iambmelt commented on testapps/testapp/src/main/AndroidManifest.xml:
```diff
@@ -55,14 +55,15 @@
         <meta-data
             android:name="com.microsoft.identity.client.ClientId"
-            android:value="9851987a-55e5-46e2-8d70-75f8dc060f21"/>
+            android:value="b92e0ba5-f86e-4411-8e18-6b5f928d968a"/>
```
Intentional?

iambmelt commented on testapps/testapp/src/main/java/com/microsoft/identity/client/testapp/AcquireTokenFragment.java:
```diff
@@ -37,6 +37,9 @@
 import android.widget.Switch;
 import android.widget.TextView;
 
+import com.google.gson.Gson;
```
Unused imports being added to this file

iambmelt commented on testapps/testapp/src/main/res/raw/msal_config_webview.json:
```diff
@@ -1,7 +1,7 @@
 {
-  "client_id" : "9851987a-55e5-46e2-8d70-75f8dc060f21",
+  "client_id" : "2523b71d-3ac3-4aea-833d-7096286ec643",
```
Intentional?

heidijinxujia replied:
Could change back after Gladwin updates the redirectUri

heidijinxujia commented on testapps/testapp/src/main/res/raw/msal_config.json:
```diff
@@ -1,7 +1,7 @@
 {
-  "client_id" : "9851987a-55e5-46e2-8d70-75f8dc060f21",
+  "client_id" : "b92e0ba5-f86e-4411-8e18-6b5f928d968a",
```
9851987a-55e5-46e2-8d70-75f8dc060f21  is not lab app client id. Going to use 4b0db8c2-9f26-4417-8bde-3f0e3656f8e0 instead.

heidijinxujia commented:
Thank you, Brian! I've addressed all your comments.
1. `9851987a-55e5-46e2-8d70-75f8dc060f21` is not a lab app. Could not find it anywhere under my subscriptions. So I update the client id to lab's as well as the redirect uri.

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -603,7 +590,7 @@ public void getAccount(@NonNull final String oid,
 
     private void getBrokerAndLocalAccount(@NonNull final String oid,
-                                          @NonNull final AccountLoadedCallback callback) {
+                                          @NonNull final AccountsLoadedCallback callback) {
```
This should generate a warning for using a raw type - will this only ever return 1?

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -554,51 +555,128 @@ public void run() {
-                    homeAccountIdentifier,
-                    realm
-            );
+        final Handler handler;
```
`getAccounts()` and `getBrokerLocalAccounts()` have the same piece of logic - let's split this into a convenience function like `getActiveLooper()` or `getPreferredLooper()`

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -554,51 +555,128 @@ public void run() {
+                                handler.post(new Runnable() {
+                                    @Override
+                                    public void run() {
+                                        callback.onAccountsLoaded(AccountAdapter.adapt(accountRecord));
```
This callback is being invoked inside a loop, surely you mean to call this only once, right?

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -554,51 +555,128 @@ public void run() {
+                        "Found the matched account."
+                );
+
+                callback.onAccountsLoaded(AccountAdapter.adapt(accountRecord));
```
Same here - callback invocation inside of a loop

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -554,51 +557,127 @@ public void run() {
+                                @Override
+                                public void run() {
+                                    //Return null if no account found.
+                                    callback.onAccountsLoaded(null);
```
Let's return an empty `List` instead - otherwise we force callers to check for both `null` and `isEmpty()`

heidijinxujia replied:
This function is trying to find the target account, the one with matching oid, from local msal and broker's cache. Not the entire account list.

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -384,23 +385,25 @@ public PublicClientApplicationConfiguration getConfiguration() {
         void onAccountsLoaded(List<AccountRecord> accountRecords);
     }
 
+    private static Handler getPreferredLooper() {
```
We can decorate this method's return type as `@NonNull`

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -473,7 +476,7 @@ public void onMigrationFinished(int numberOfAccountsMigrated) {
      * @param callback: AccountsLoadedCallback
      */
-    private void postLocalAccountsResult(final Handler handler, final AccountsLoadedCallback callback) {
+    private void postLocalAccountsResult(final Handler handler, final AccountsLoadedCallback<List<IAccount>> callback) {
```
Both params here `@NonNull`

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -493,7 +496,7 @@ public void run() {
      * @param callback: AccountsLoadedCallback
      */
-    private void postBrokerAndLocalAccountsResult(final Handler handler, final AccountsLoadedCallback callback) {
+    private void postBrokerAndLocalAccountsResult(final Handler handler, final AccountsLoadedCallback<List<IAccount>> callback) {
```
Both params `@NonNull`

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -554,51 +557,127 @@ public void run() {
+                    TAG + methodName,
+                    "Eligible to use broker. Try to get the account from local and broker cache."
+            );
+            getBrokerAndLocalAccount(oid, callback);
```
`getBrokerAndLocalAccounts()` (plural)

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -554,51 +557,127 @@ public void run() {
-    @Nullable
-    public IAccount getAccount(@NonNull final String homeAccountIdentifier,
-                               @Nullable final String authority) {
+    public void getAccount(@NonNull final String oid,
```
I think that the factoring of this method is more complicated than it needs to be....

```java
void getAccount(@NonNull final String oid, @NonNull final String AccountsLoadedCallback<IAccount> callback) {
    final AccountsLoadedCallback<IAccount> brokerCallback = new AccountsLoadedCallback<>() {
        @Override
        public void onAccountsLoaded(@Nullable final IAccount account) {
            callback.onAccountsLoaded(account);
        }
    }

    getLocalAccountByOid(oid, new AccountsLoadedCallback<IAccount>() {
        @Override
        public void onAccountsLoaded(@Nullable final IAccount account) {
            if (null == account) {
                getBrokerAccountByOid(oid, brokerCallback); // Method yields null if not broker elligible
            } else {
                callback.onAccountsLoaded(account);
            }
        }
    })
}
```

Then.... we define two methods...

```java
// One method to load accounts locally and return a List
private void getLocalAccountByOid(@NonNull String oid, @NonNull final AccountsLoadedCallback<IAccount> callback) {
    // On a worker thread, if necessary....
    final List<IAccount> accounts = getLocalAccounts();
    callback.onAccountsLoaded(getAccountByOid(acounts));
}
````

````java
// Another method to load Accounts from Broker and return a List
private void getBrokerAccountByOid(@NonNull String oid, @NonNull final AccountsLoadedCallback<IAccount> callback) {
    // On a worker thread, if necessary....
    final List<IAccount> accounts = getBrokerAccounts(); // null or empty, as appropriate (null for no-broker, empty for no-accounts)
    callback.onAccountsLoaded(getAccountByOid(accounts));
}
````

And getAccountByOid() does all the filtering once....

```java
public IAccount getAccountByOid(@NonNull final String oid, @NonNull List<IAccount> accounts) {
    // fill in the rest here....
}
````

heidijinxujia replied:
I'm afraid the getBrokerAccountByOid cannot be refactored in this way. Because the call to broker cannot happen on main thread like the getLocalAccount, so it has to call the looper to post the result on a different thread. Thus, it is better to get both broker and local accounts on the same thread, fetch the target account and return it back.

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -554,51 +557,127 @@ public void run() {
+                        }
+
+                        for (final AccountRecord accountRecord : cachedAccounts) {
+                            if (!StringUtil.isEmpty(accountRecord.getLocalAccountId())
```
This loop and its conditions are duplicated across methods [here ](https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/8709d0d6214f9e67a778096fcf102801752bc8ff/msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java#L626-L643)and [here](https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/8709d0d6214f9e67a778096fcf102801752bc8ff/msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java#L663-L673) -- we should split this into a few different functions that we can call again and again, rather than duplicate --[ I have specific suggestions here](https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/592/files#r281266640).

heidijinxujia commented:
Comments addressed. Please have a look. Thanks.

iambmelt commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -555,51 +561,104 @@ public void run() {
+    }
+
+    private IAccount getAccountByOid(@NonNull final String oid,
+                                     List<AccountRecord> accounts) {
```
Minor nit here on `final` and `@NonNull`

iambmelt commented on testapps/testapp/src/main/java/com/microsoft/identity/client/testapp/MainActivity.java:
```diff
@@ -269,21 +269,11 @@ public PublicClientApplication getPublicClientApplication() {
 
-        //final IAccount requestAccount = getAccount();
-        mApplication.getAccounts(new PublicClientApplication.AccountsLoadedCallback() {
+        mApplication.getAccount(requestOptions.getLoginHint().trim(), new PublicClientApplication.AccountsLoadedCallback<IAccount>() {
```
Why is the `loginHint` being passed here?

heidijinxujia replied:
This is the loginhint text box in our test app. We can pass in the oid for testing here. I will add a comment there.

shoatman commented on msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:
```diff
@@ -555,51 +561,104 @@ public void run() {
-    @Nullable
-    public IAccount getAccount(@NonNull final String homeAccountIdentifier,
-                               @Nullable final String authority) {
+    public void getAccount(@NonNull final String oid,
```
I believe that we'll need to support getting an account by OID, HomeAccountId, etc... so let's make that possible by accepting a parameter object.

heidijinxujia commented:
This PR is out of sync after AccountManager change. Close this one and open another PR for it.
